datafusion/common/src/tree_node.rs

```diff
@@ -532,6 +669,25 @@ impl<T> Transformed<T> {
     }
 }
 
+impl Transformed<()> {
+    /// Invoke the given function `f`  and combine the transformed state with
+    /// the current state:
+    ///
+    /// * if `f` returns an Err, returns that err
+    ///
+    /// * If `f` returns Ok, sets `self.transformed` to `true` if either self or
+    /// the result of `f` were transformed.
+    pub fn and_then<F>(self, f: F) -> Result<Transformed<()>>
```

Comment:
Introducing this function allowed the implementation of mutate and mutate children to be more concise. I couldn't figure out how to implement it in general for two `Transformed` as there are two different payloads